vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,318 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+use actor::{ActorState, ACCOUNT_ACTOR_CODE_ID, REWARD_ACTOR_ADDR};
 use address::Address;
 use blocks::Tipset;
 use clock::ChainEpoch;
-use default_runtime::{internal_send, DefaultRuntime};
+use default_runtime::{
+    internal_send, transfer_from_gas_holder, transfer_to_gas_holder, DefaultRuntime,
+};
+use forest_encoding::Cbor;
 use ipld_blockstore::BlockStore;
+use log::warn;
 use message::{Message, MessageReceipt, SignedMessage, UnsignedMessage};
 use num_bigint::BigUint;
-use num_traits::cast::ToPrimitive;
-use runtime::DefaultSyscalls;
-use vm::ActorError;
-use vm::{ExitCode, Serialized, StateTree};
-
+use num_traits::Zero;
+use runtime::Syscalls;
+use vm::{price_list_by_epoch, ActorError, ExitCode, Serialized, StateTree};
 /// Interpreter which handles execution of state transitioning messages and returns receipts
 /// from the vm execution.
-pub struct VM<'a, ST: StateTree, DB: BlockStore> {
+pub struct VM<'a, ST: StateTree, DB: BlockStore, SYS: Syscalls>
+where
+    SYS: Copy,
+{
     state: ST,
     store: &'a DB,
     epoch: ChainEpoch,
+    syscalls: SYS,
     // TODO: missing fields
 }
 
-impl<'a, ST: StateTree, DB: BlockStore> VM<'a, ST, DB> {
-    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch) -> Self {
+impl<'a, ST: StateTree, DB: BlockStore, SYS: Syscalls> VM<'a, ST, DB, SYS>
+where
+    SYS: Syscalls + Copy,
+{
+    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch, syscalls: SYS) -> Self {
         VM {
             state,
             store,
             epoch,
+            syscalls,
         }
     }
 
+    /// Returns ChainEpoch
+    fn epoch(&self) -> &ChainEpoch {
+        &self.epoch
+    }
+
     /// Apply all messages from a tipset
     /// Returns the receipts from the transactions.
     pub fn apply_tip_set_messages(
         &mut self,
         _tipset: &Tipset,
         msgs: &TipSetMessages,
-    ) -> Result<Vec<MessageReceipt>, String> {
+    ) -> Result<Vec<ApplyRet>, String> {
```

Comment:
Can you reference to why this extra data is needed to be returned?

Reply:
The `penalty` and `actor_err` are being used in state manager component, for example in [ApplyBlocks](https://github.com/filecoin-project/lotus/blob/testnet/3/chain/stmgr/stmgr.go#L170).

If it's determined that these fields can be derived elsewhere or they are not needed in state manager I would be happy to switch back to returning `MessageReceipt`, would look much cleaner.